File: py110-intermediate/exercises/medium_1/fib_mem.py
memo = {}  # Dictionary to store results

def fibonacci(n):
    if n <= 2:
        return 1

    if n in memo:
        return memo[n]  # Return saved result

    memo[n] = fibonacci(n - 1) + fibonacci(n - 2)  # Save result
    return memo[n]

print(fibonacci(1) == 1)         # True
print(fibonacci(2) == 1)         # True
print(fibonacci(3) == 2)         # True
print(fibonacci(4) == 3)         # True
print(fibonacci(5) == 5)         # True
print(fibonacci(6) == 8)         # True
print(fibonacci(12) == 144)      # True
print(fibonacci(20) == 6765)     # True
print(fibonacci(100))   # True

File: ls215/exercises/medium1/fibMem.js
const memory = {};

function fibonacci(n) {
  if (n <= 2) {
    memory[n] = 1;
    return memory[n];
  }

  memory[n] = fibonacci(n - 1) + memory[n - 2];

  return memory[n];
}


console.log(fibonacci(1));       // 1
console.log(fibonacci(2));       // 1
console.log(fibonacci(3));       // 2
console.log(fibonacci(4));       // 3
console.log(fibonacci(5));       // 5
console.log(fibonacci(6));       // 5
// console.log(fibonacci(12));      // 144
// console.log(fibonacci(20));      // 6765